components/style/gecko/conversions.rs

```diff
@@ -338,6 +340,182 @@ impl nsStyleImage {
+        use values::generics::image::{ColorStop, CompatMode, Circle, Ellipse, EndingShape, GradientKind, ShapeExtent};
+        use values::specified::length::Percentage;
+        use values::specified::position::{X, Y};
+
+        let gecko_gradient = Gecko_GetGradientImageValue(self).as_ref().unwrap();
+        let gradient_kind = match gecko_gradient.mShape as u32 {
+            NS_STYLE_GRADIENT_SHAPE_LINEAR => {
+                let line_direction = match *gecko_gradient.mAngle.get_mUnit() {
+                    eStyleUnit_None => {
+                        debug_assert!(*gecko_gradient.mBgPosX.get_mUnit() == eStyleUnit_Percent &&
+                                      *gecko_gradient.mBgPosY.get_mUnit() == eStyleUnit_Percent);
+                        let pos_x = LengthOrPercentage::from_gecko_style_coord(&gecko_gradient.mBgPosX)
+                            .expect("mBgPosX has valid data");
+                        let pos_y = LengthOrPercentage::from_gecko_style_coord(&gecko_gradient.mBgPosY)
+                            .expect("mBgPosY has valid data");
+                        let horizontal_direction = match pos_x {
+                            LengthOrPercentage::Percentage(Percentage(0.0)) => X::Left,
+                            LengthOrPercentage::Percentage(Percentage(1.0)) => X::Right,
+                            x => panic!("Found unexpected mBgPosX in nsStyleImage: {:?}", x),
+                        };
+                        let vertical_direction = match pos_y {
+                            LengthOrPercentage::Percentage(Percentage(0.0)) => Y::Top,
+                            LengthOrPercentage::Percentage(Percentage(1.0)) => Y::Bottom,
+                            x => panic!("Found unexpected mBgPosY in nsStyleImage: {:?}", x),
+                        };
+                        LineDirection::Corner(horizontal_direction, vertical_direction)
+                    },
+                    _ => {
+                        LineDirection::Angle(Angle::from_gecko_style_coord(&gecko_gradient.mAngle)
+                                             .expect("gecko_gradient.mAngle has valid data"))
+                    }
+                };
+                GradientKind::Linear(line_direction)
+            },
+            _ => {
+                let gecko_size_to_keyword = |gecko_size| {
+                    match gecko_size {
+                        NS_STYLE_GRADIENT_SIZE_CLOSEST_SIDE => ShapeExtent::ClosestSide,
+                        NS_STYLE_GRADIENT_SIZE_FARTHEST_SIDE => ShapeExtent::FarthestSide,
+                        NS_STYLE_GRADIENT_SIZE_CLOSEST_CORNER => ShapeExtent::ClosestCorner,
+                        NS_STYLE_GRADIENT_SIZE_FARTHEST_CORNER => ShapeExtent::FarthestCorner,
+                        // We don't choose ShapeExtent::Contain and ShapeExtent::Cover
+                        // since those 'contain' and 'cover' keywords are handled as synonyms of the standard
```

Comment:
Please let me fix this after fixing https://bugzilla.mozilla.org/show_bug.cgi?id=1217664
I'll write a memo in that source code.